Add ability to specify subprojects

brandonio21 opened the pull request:
### Problem

Including a subproject (with its own build root and own build-root-relative BUILD targets) within a project and depending on targets of the subproject was previously unsupported. All subproject BUILD targets would not resolve properly since they would be relative to the subproject's buildroot, not the main project's build root. 

Further discussion found in issue https://github.com/pantsbuild/pants/issues/3978

### Solution

Add a new configuration option, `subproject-roots`, that allows the user to specify roots of all subprojects. Upon resolving targets in a BUILD file, the BUILD file path is compared to subproject roots. If the BUILD file resides in a subproject, then the target is corrected (The relative path between the project root and the subproject root is prepended to the spec).

### Result

It is now possible to include subprojects in larger projects. This PR is only limited in scope; however. The subproject's targets are still built using the main project's configurations, etc. In the future, it may be a nicety to allow the subproject's targets to be built with the subproject's `pants` and `pants.ini` files.

baroquebobcat commented:
Could you add tests that exercise the new behavior?

stuhood commented:
Thanks @brandonio21 ! We just switched to github PRs as review process: if this is ready for review, please see http://www.pantsbuild.org/howto_contribute.html , and in particular, add the `reviewable` label.

brandonio21 commented:
@stuhood @baroquebobcat 

Sure thing - I'll go ahead and add some tests and add the label as soon as I get the chance :)

brandonio21 commented:
I've added some unit tests for the core PR functionality. Also, as an FYI, those who do not have write access to the repo may not add labels :(

stuhood commented:
Blargh... thanks for pointing out the label thing. Will follow up.

stuhood commented on src/python/pants/util/dirutil.py:
```diff
@@ -19,6 +19,23 @@
 from pants.util.strutil import ensure_text
 
 
+def longest_dir_prefix(path, prefixes):
```
This is the kind of thing that would be easiest accomplished with the Trie that is used for source_roots.

benjyw replied:
It might (or might not) be more efficient to use a trie (which would have to be refactored out of the source roots code), but I don't mind if that happens in a followup change, and only if there's a noticeable performance gain. The SourceRootTrie code splits the paths, generating small temporary strings that then need to be GC'd, whereas I assume `str.startswith` is highly optimized in the python interpreter, so this simple implementation might actually be faster in practice, for small numbers of prefixes...

stuhood commented:
While the new test is very useful, I worry that this is not hooked up completely from end-to-end. In particular, an integration test with a subproject is probably a blocker for landing this.

I'd recommend creating an integration test like: https://github.com/pantsbuild/pants/blob/f3f6f012f7f43c2f4632e12b103362dd399b4560/tests/python/pants_test/engine/legacy/test_changed_integration.py ... and create a tiny anonymous subdirectory below another working copy.

Two potential approaches for the test:
1. Create a subproject under the `pantsbuild/pants` repo (say, under `testprojects`) and then have the integration test use options that declare the project temporarily. You might possibly need to ignore the BUILD files in this subdirectory.
2. Have the integration test create two anonymous "repos" in the test, nested within one another.

benjyw commented:
Hi all,

I'm on vacation with almost no internet access, but I'd like to take a
deeper look at this, to see what these "subprojects" are, and how they
interact with source roots. So please don't land this until I have a chance
to do so, probably on Monday Dec. 5th.

Kia ora,

Benjy

On Tue, Nov 29, 2016 at 12:39 PM Stu Hood <notifications@github.com> wrote:

> While the new test is very useful, I worry that this is not hooked up
> completely from end-to-end. In particular, an integration test with a
> subproject is probably a blocker for landing this.
>
> I'd recommend creating an integration test like:
> https://github.com/pantsbuild/pants/blob/f3f6f012f7f43c2f4632e12b103362dd399b4560/tests/python/pants_test/engine/legacy/test_changed_integration.py
> ... and create a tiny anonymous subdirectory below another working copy.
>
> Two potential approaches for the test:
>
>    1. Create a subproject under the pantsbuild/pants repo (say, under
>    testprojects) and then have the integration test use options that
>    declare the project temporarily. You might possibly need to ignore the
>    BUILD files in this subdirectory.
>    2. Have the integration test create two anonymous "repos" in the test,
>    nested within one another.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pantsbuild/pants/pull/4088#issuecomment-263430445>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAfS_Mn7kidDL5NXSpZlwrPVxuRXvhr8ks5rC2YtgaJpZM4K3rVF>
> .
>

benjyw commented on tests/python/pants_test/build_graph/test_build_file_address_mapper.py:
```diff
@@ -64,6 +64,29 @@ def test_scan_addresses_with_invalid_root(self):
 
+  def test_determine_subproject_spec(self):
+    # Ensure that a spec referring to a subproject gets assigned to that
+    # subproject properly
```
Grammar nit: Please end all sentences with periods (despite bad examples elsewhere in this file...).

benjyw commented on tests/python/pants_test/util/test_dirutil.py:
```diff
@@ -31,6 +31,23 @@ def setUp(self):
     _mkdtemp_unregister_cleaner()
 
+  def test_longest_dir_prefix(self):
+    # Find the longest prefix (standard case)
```
Nit: Period.

benjyw commented on src/python/pants/build_graph/build_file_address_mapper.py:
```diff
@@ -102,6 +104,15 @@ def addresses_in_spec_path(self, spec_path):
+    if subproject_prefix:
+      spec = os.path.join(subproject_prefix, spec)
+
+    logger.debug("Determined that spec {} relative to {} belongs to subproject "
```
Nit: prefer single quotes for strings.

stuhood commented:
Hey @brandonio21! Do you have time to address @benjyw 's feedback? If so, we should be able to land this!

brandonio21 commented:
Sorry for the delay on everything. I decided to work on this feature right before I went back to school. Turns out that wasn't the best idea in terms of speedy PR merging.

For whatever reason, pants is no longer working for me (Something about `Could not satisfy all requirements for twitter.common.lang==0.3.7`), so I cannot run the new tests on my system at the moment (Also don't have much time to debug).

I have squashed my changes and added another test. Let's see what Travis thinks. I will use him as my tester for now.

stuhood commented:
@brandonio21 : It looks like there are some issues with the tests...

To get your local copy of pants running again, you might need to clear the pex cache? Try `git clean -fdx`, or, failing that, `rm ~/.pex` to clear the pex cache.

brandonio21 commented:
Alright, I've gone ahead and done the following:

- Add an integration test which tests an artificial subproject in `testprojects/src/python/subproject_test`
- Add `join_specs` function to dirutil. This function is a result of the pre-pended "//" in the subproject's dependency listed in their BUILD file. 
- Add a unit test for aforementioned `join_specs`

Things should now be good to go! :)

brandonio21 commented:
@stuhood All is good to go! There were a few changes made and an integration test added. Let me know your thoughts! 

Note that having the BUILD files for a broken dependency due to lack of "subproject" flag would result in failures in a bunch of other tests. So I created the files on demand. It is a working solution, but do let me know if it is not acceptable.
